python/tvm/contrib/pipeline_executor.py

```diff
@@ -0,0 +1,559 @@
+                Module information like index, module name
+
+            io_type : str
+                The type of this binding. It can be either "input" or "output".
+
+            name : str/integer
+                Binding name, for input that is string for example "data0"
+                for output that is integer for example 0.
+            """
+
+            def __init__(self, owner, stype, name, data_type=None):
+                self.io_owner = owner
+                self.io_type = stype
+                self.name = str(name)
+                # These item that have dependency relation with self
+                self.bindings = []
+                # The item that self depend
+                self.parents = []
+
+                self.data_type = data_type
+
+            def get_name(self):
+                """get owner name and self name"""
+                owner_name = ""
+                if isinstance(self.io_owner, PipelineConfig.ModuleWrapper):
+                    owner_name = self.io_owner.name
+
+                return owner_name, self.name
+
+            def get_owner_indx(self):
+                """return index if owner is ModuleWrapper, if not return 0"""
+                if isinstance(self.io_owner, PipelineConfig.ModuleWrapper):
+                    return self.io_owner.index
+
+                # if not ModuleWrapper then owner is PipelineConfig, return 0
+                # to identify this is global interface
+                return 0
+
+            def __repr__(self):
+                """Get all binding(input data), exepect like |data_0: mod1:data_0"""
+                ret = "  |{}: ".format(self.name)
+                for binding in self.bindings:
+                    mname, dname = binding.get_name()
+                    ret += "{0}:{1} ".format(mname, dname)
+                return ret
+
+            def dag_acircle_check(self, start, inputs):
+                """check if the DAG that current binding stay is acircle"""
+                for _, binding in inputs.items():
+                    if start == binding.io_owner:
+                        return False
+                    for p in binding.parents:
+                        if not self.dag_acircle_check(start, p.io_owner.input_bindings.bindings):
+                            return False
+
+                return True
+
+            def connect(self, binding):
+                """
+                # check if the bindendency setting correct.
+                # correct connection are following
+                # 1. global input to module input
+                # 2. module output to global output
+                # 3. module output to moudle input
+                """
+                owner_indx = self.get_owner_indx()
+                bind_owner_indx = binding.get_owner_indx()
+                if owner_indx == bind_owner_indx:
+                    raise RuntimeError(f"can not set self as binding.")
```

Comment:
By getting rid of the index, you should implement this check like
```
if self.io_owner == binding.io_owner:
    raise RuntimeError(f"Can not set self as binding.")
```

Then you need to implement `ModuleWrapper.__eq__(self, other)` to support the `==` operator.

Reply:
fixed.